### PR TITLE
Add matplotlib plotting function

### PR DESCRIPTION
**Pull request by peterfpeterson**
To improve re-use between users and the workbench. Add "simple" wrappers that unpack our data objects for consumption by matplotlib. These should be reusable by mantid and do as little work with matplotlib as possible.

This has #21232 merged into it already. It is an unrelated PR, but should be merged into `master` before this one.

**To test:**

What I was using (but dataset can be replaced):
```python
from mantid.simpleapi import mtd, LoadEventNexus, Rebin, ConvertUnits
from mantid import plots

LoadEventNexus(Filename="PG3_12345", OutputWorkspace="PG3_12345")
Rebin(InputWorkspace="PG3_12345", OutputWorkspace="PG3_12345", Params=100)

import matplotlib.pyplot as plt
fig, ax = plt.subplots()
plots.plot(ax, mtd['PG3_12345'], 'go-', specNum=2500, label='user label')
plots.errorbar(ax, mtd['PG3_12345'], specNum=2501)
ax.legend()
fig.show()

ConvertUnits(InputWorkspace='PG3_12345', OutputWorkspace='PG3_12345', Target='dSpacing')

fig, ax = plt.subplots()
for i in range(2000,2500):
    plots.plot(ax, mtd['PG3_12345'], specNum=i)
fig.show()



fig, ax = plt.subplots()
c = plots.contourf(ax, mtd['PG3_12345'])
ax.set_xlabel('awesome label')
fig.colorbar(c)
fig.show()

LoadEventNexus(Filename='CNCS_7860_event.nxs', OutputWorkspace='CNCS_7860_event')
ConvertUnits(InputWorkspace='CNCS_7860_event', OutputWorkspace='CNCS_7860_event', Target='DeltaE', EMode='Direct', EFixed=3)
Rebin(InputWorkspace='CNCS_7860_event', OutputWorkspace='CNCS_7860_event', Params='-3,0.05,3')
SofQW(InputWorkspace='CNCS_7860_event', OutputWorkspace='CNCS_7860_sqw', QAxisBinning='0,0.05,3', EMode='Direct', EFixed=3)

print(plots.getAxesLabels(mtd['PG3_12345']))

fig, ax = plt.subplots()
c = plots.contourf(ax, mtd['CNCS_7860_sqw'])
ax.set_xlabel('awesome label')
fig.colorbar(c)
fig.show()

fig, ax = plt.subplots()
c = plots.pcolormesh(ax, mtd['CNCS_7860_sqw'])
fig.colorbar(c)
fig.show()
```

*There is no associated issue.*

**Release Notes** 

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by peterfpeterson** [comment]
Comparison of "mantidplot" figure and this. The matplotlib is intentionally ugly to expose how the user would specify some of the customization.

pg3_12345_mantid
pg3_12345_mpl

File: qt/python/mantidqt/plots/pyplot.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
+def plot(axes, workspace, *args, **kwargs): #specNum=None, wkspIndex=None, distribution=None, *args, **kwargs):
```
Comment (martyngigg) [inline qt/python/mantidqt/plots/pyplot.py]: This looks really nice.

My only concern here is that as we've labelled the module `pyplot` then I think it should be possible to get a plot with a single line as in `mpl.pyplot`. I think users will quickly get irritated if they always have to create the figures and axes themselves. Can we make them optional and use `matplotlib.pyplot.gca()` to create one if we're not given on?

Comment (peterfpeterson) [reply qt/python/mantidqt/plots/pyplot.py]: I don't want to encourage users to use the stateful matplotlib API. It creates all sorts of issues. I'd rather rename this module to better communicate what it is.

Comment (martyngigg) [reply qt/python/mantidqt/plots/pyplot.py]: How about `mantidqt.plots.workspace`?

I see your point for scripts - it's much easier to deal with manipulating existing plots this way.
I think we will rapidly be asked for one liners for the common plot types though. Especially if you want to do something quickly from the IPython terminal.

The other question is what about the custom toolbars for 1D & 2D plots that we had designed - how would we deal with those?

**Comment by peterfpeterson** [comment]
Rename package to just `mantid.plots` and add functionality for error bars. Additional functions for `contour` and `contourf` will be added. `mantid.plots.pyplot` will be added as a future PR to reproduce functionality that already exists in `pymantidplot.pyplot` using matplotlib.

File: Framework/PythonInterface/mantid/plots/_functions.py [hunk 2]
```diff
@@ -0,0 +1,188 @@
+import numpy
```
Comment (martyngigg) [inline Framework/PythonInterface/mantid/plots/_functions.py]: Can we add the GPL licence header and a module doc string?

Comment (peterfpeterson) [reply Framework/PythonInterface/mantid/plots/_functions.py]: The module doc string is in the `__init__.py` file.

**Comment by peterfpeterson** [comment]
This is ready to be tested. The real examples are in the module documentation.

File: Framework/PythonInterface/mantid/CMakeLists.txt [hunk 3]
```diff
@@ -35,35 +35,36 @@ add_custom_target ( PythonModule DEPENDS ${PYTHON_PKGROOT_INSTALL_FILES} ${PYTHO
...
 add_subdirectory ( dataobjects )
 add_subdirectory ( _plugins )
+add_subdirectory ( plots )
 
 # Create an overall target
-add_custom_target ( PythonInterface DEPENDS PythonKernelModule PythonGeometryModule 
-                    PythonAPIModule PythonDataObjectsModule PythonCurveFittingModule )
+add_custom_target ( PythonInterface DEPENDS PythonKernelModule PythonGeometryModule
+                    PythonAPIModule PythonDataObjectsModule PythonCurveFittingModule
+                    PythonPlotsModule)
```
Comment (martyngigg) [inline Framework/PythonInterface/mantid/CMakeLists.txt]: Can you change the framework (CMakeLists.txt)[/mantidproject/mantid/blob/master/Framework/CMakeLists.txt#L147] and remove the `Python*` targets and replace it with `PythonInterface`. This should ensure a `make Framework` gets everything.

File: docs/source/api/python/mantid/plots/index.rst [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+.. figure:: ../../../../images/mantid_plots_contourf.png
+   :align: center
+   :figwidth: image
+
+Similarly, showing the actual values with :func:`~mantid.plots.pcolormesh`
+
+.. code-block:: python
+
+   fig, ax = plt.subplots()
+   c = plots.pcolormesh(ax, mtd['CNCS_7860_sqw'])
```
Comment (martyngigg) [inline docs/source/api/python/mantid/plots/index.rst]: Could we find a prettier picture for these two?

**Review by martyngigg — CHANGES_REQUESTED** [review]

**Review by AndreiSavici — APPROVED** [review]
Looks fine to me

**Review by martyngigg — APPROVED** [review]

**Pull request merged**